File: vol1/ch10/LeetCode399.py
class Solution:
    def calcEquation(self,equations:List[List[str]],values:List[float],queries:List[List[str]]) -> List[float]:
        m,n=len(queries),len(equations)
        self.Init(2*n)
        id=0                           #给不同变量从0开始编号
        hmap={}
        for i in range(0,n):
            if equations[i][0] not in hmap:
                hmap[equations[i][0]]=id;id+=1
            if equations[i][1] not in hmap:
                hmap[equations[i][1]]=id;id+=1
            x=hmap[equations[i][0]]
            y=hmap[equations[i][1]]
            self.Union(x,y,values[i])
        ans=[None]*m;              #存放答案
        for i in range(0,m):
            if queries[i][0] not in hmap or queries[i][1] not in hmap:
                ans[i]=-1.0
                continue
            x=hmap[queries[i][0]]
            y=hmap[queries[i][1]]
            rx=self.Find(x)
            ry=self.Find(y)
            if rx==ry: ans[i]=self.weight[x]/self.weight[y]
            else: ans[i]=-1.0
        return ans

    def Init(self,n):
        self.parent=[None]*n;
        self.weight=[None]*n;
        for i in range(0,n):
            self.parent[i]=i
            self.weight[i]=1.0

    def Find(self,x):
        if x!=self.parent[x]:
            origin=self.parent[x]
            self.parent[x]=self.Find(self.parent[x])
            self.weight[x]*=self.weight[origin]
        return self.parent[x]

    def Union(self,x,y,val):
        rx=self.Find(x)
        ry=self.Find(y)
        if rx==ry:return
        self.parent[rx]=ry
        self.weight[rx]=val*self.weight[y]/self.weight[x]

File: vol1/ch10/LeetCode399.cpp
class Solution {
    vector<int> parent;
    vector<double> weight;
public:
    vector<double> calcEquation(vector<vector<string>>& equations,vector<double>&values, vector<vector<string>>&queries) {
        int n=equations.size();
        Init(2*n);
        int id=0;                           //给不同变量从0开始编号
        unordered_map<string,int> hmap;
        int x,y;
        for(int i=0;i<n;i++) {
            if(hmap.count(equations[i][0])==0)
                hmap[equations[i][0]]=id++;
            if(hmap.count(equations[i][1])==0)
                hmap[equations[i][1]]=id++;
            x=hmap[equations[i][0]];
            y=hmap[equations[i][1]];
            Union(x,y,values[i]);
        }
        int m=queries.size();
        vector<double> ans(m);              //存放答案
        for(int i=0;i<m;i++) {
            if(hmap.count(queries[i][0])==0 || hmap.count(queries[i][1])==0) {
                ans[i]=-1.0;
                continue;
            }
            x=hmap[queries[i][0]];
            y=hmap[queries[i][1]];
            int rx=Find(x);
            int ry=Find(y);
            if (rx==ry) ans[i]=weight[x]/weight[y];
            else ans[i]=-1.0;
        }
        return ans;
    }
    void Init(int n) {
        parent=vector<int>(n);
        weight=vector<double>(n);
        for (int i=0;i<n;i++) {
            parent[i]=i;
            weight[i]=1.0;
        }
    }
    int Find(int x) {
        if(x!=parent[x]) {
            int origin=parent[x];
            parent[x]=Find(parent[x]);
            weight[x]*=weight[origin];
        }
        return parent[x];
    }

    void Union(int x,int y,double val){
        int rx=Find(x);
        int ry=Find(y);
        if(rx==ry) return;
        parent[rx]=ry;
        weight[rx]=val*weight[y]/weight[x];
    }
};

